File: blendertools/CPCWMap_Blender/protodb.py
"""Minimal ProtoDB.bin reader (stdlib only, Blender-safe).

Vendored from cpcw_protodb.py — just enough to build a GUID -> model-path index
so the map importer can resolve each entity's Prototype to its .srm model.
"""

import struct

FT_INT32 = 0x0001
FT_FLOAT = 0x0002
FT_BOOL = 0x0003
FT_STRING = 0x0004
FT_GUID = 0x0011
FT_REF = 0x0012
FT_UINT8 = 0x0017
FT_INT16 = 0x0019
FT_INLINE1 = 0x0088
FT_INLINE2 = 0x0089
FT_FLAGS = 0x039c
FT_BLOB = 0x0165
FT_ARRAY = 0x898a


class Schema:
    __slots__ = ('name', 'type_id', 'version', 'fields')

    def __init__(self, name, type_id, version, fields):
        self.name = name
        self.type_id = type_id
        self.version = version
        self.fields = fields


class ProtoDB:
    def __init__(self, filepath):
        with open(filepath, 'rb') as f:
            self.data = f.read()
        self.schemas = {}
        self._parse_header()
        self._parse_schemas()

    def _parse_header(self):
        if self.data[0:4] != b'OBJS':
            raise ValueError("Not a ProtoDB file (expected OBJS)")
        self.total_size = struct.unpack_from('<I', self.data, 4)[0]
        self.schema_offset = struct.unpack_from('<I', self.data, 8)[0]

    def _parse_schemas(self):
        pos = self.schema_offset
        if self.data[pos:pos + 4] != b'SCHD':
            raise ValueError("Expected SCHD at 0x%x" % pos)
        pos += 8
        schema_count = struct.unpack_from('<H', self.data, pos)[0]
        pos += 4
        for _ in range(schema_count):
            if pos + 8 > len(self.data) or self.data[pos:pos + 4] != b'SCHM':
                break
            content_sz = struct.unpack_from('<I', self.data, pos + 4)[0]
            cs = pos + 8
            ce = cs + content_sz
            p = cs
            name_len = struct.unpack_from('<H', self.data, p)[0]; p += 2
            name = self.data[p:p + name_len].decode('ascii', 'replace'); p += name_len
            type_id, version, field_count = struct.unpack_from('<HHH', self.data, p); p += 6
            fields = []
            for _ in range(field_count):
                if p + 2 > ce:
                    break
                fn_len = struct.unpack_from('<H', self.data, p)[0]; p += 2
                if p + fn_len > ce:
                    break
                fn = self.data[p:p + fn_len].decode('ascii', 'replace'); p += fn_len
                if p + 8 > ce:
                    break
                ftype, fsize = struct.unpack_from('<II', self.data, p); p += 8
                fields.append((fn, ftype, fsize))
            self.schemas[type_id] = Schema(name, type_id, version, fields)
            pos = ce

    def parse_root(self):
        obj, _ = self._parse_objt(12)
        return obj

    def _parse_objt(self, pos):
        if self.data[pos:pos + 4] != b'OBJT':
            return None, pos
        content_sz = struct.unpack_from('<I', self.data, pos + 4)[0]
        type_id = struct.unpack_from('<H', self.data, pos + 8)[0]
        content_end = pos + 8 + content_sz
        obj, vobj_end = self._parse_vobj(pos + 10)
        if obj is None:
            obj = {'_type_id': type_id}
        p = vobj_end
        while p + 10 <= content_end and self.data[p:p + 4] == b'VOBJ':
            trailing, p = self._parse_vobj(p)
            if trailing:
                for k, v in trailing.items():
                    if not k.startswith('_'):
                        obj[k] = v
        return obj, content_end

    def _parse_vobj(self, pos):
        if self.data[pos:pos + 4] != b'VOBJ':
            return None, pos
        content_sz = struct.unpack_from('<I', self.data, pos + 4)[0]
        type_id = struct.unpack_from('<H', self.data, pos + 8)[0]
        content_end = pos + 8 + content_sz
        schema = self.schemas.get(type_id)
        obj = {'_type_id': type_id,
               '_type': schema.name if schema else 'Unknown_0x%04x' % type_id}
        p = pos + 10
        if p + 2 > content_end:
            return obj, content_end
        p += 2  # version
        if schema:
            for fname, ftype, fsize in schema.fields:
                if p >= content_end:
                    break
                val, p = self._read_field(p, ftype, fsize, content_end)
                obj[fname] = val
        return obj, content_end

    def _read_field(self, pos, ftype, fsize, limit):
        if pos >= limit:
            return None, pos
        if ftype == FT_INT32:
            return (struct.unpack_from('<i', self.data, pos)[0], pos + 4) if pos + 4 <= limit else (None, limit)
        elif ftype == FT_FLOAT:
            return (round(struct.unpack_from('<f', self.data, pos)[0], 6), pos + 4) if pos + 4 <= limit else (None, limit)
        elif ftype == FT_BOOL:
            return (bool(self.data[pos]), pos + 1) if pos + 1 <= limit else (None, limit)
        elif ftype in (FT_STRING, FT_GUID, FT_REF, FT_FLAGS):
            if pos + 2 > limit:
                return None, limit
            slen = struct.unpack_from('<H', self.data, pos)[0]; pos += 2
            if slen == 0:
                return '', pos
            end = min(pos + slen, limit)
            return self.data[pos:end].decode('ascii', 'replace'), end
        elif ftype == FT_UINT8:
            return (self.data[pos], pos + 1) if pos + 1 <= limit else (None, limit)
        elif ftype == FT_INT16:
            return (struct.unpack_from('<h', self.data, pos)[0], pos + 2) if pos + 2 <= limit else (None, limit)
        elif ftype == FT_ARRAY:
            return self._read_array(pos, limit)
        elif ftype in (FT_INLINE1, FT_INLINE2):
            return self._read_inline_object(pos, limit)
        elif ftype == FT_BLOB:
            nbytes = fsize if 0 < fsize != 65535 else 0
            if nbytes and pos + nbytes <= limit:
                return self.data[pos:pos + nbytes].hex(), pos + nbytes
            return None, pos
        else:
            if 0 < fsize != 65535 and pos + fsize <= limit:
                return None, pos + fsize
            return None, pos

    def _read_array(self, pos, limit):
        if pos + 12 > limit or self.data[pos:pos + 4] != b'ARRY':
            return [], pos
        arr_size = struct.unpack_from('<I', self.data, pos + 4)[0]
        arr_count = struct.unpack_from('<I', self.data, pos + 8)[0]
        arr_end = pos + 8 + arr_size
        items = []
        p = pos + 12
        for _ in range(arr_count):
            if p >= arr_end or self.data[p:p + 4] != b'OBJT':
                break
            obj, p = self._parse_objt(p)
            if obj:
                items.append(obj)
        return items, arr_end

    def _read_inline_object(self, pos, limit):
        if pos + 8 > limit:
            return None, pos
        tag = self.data[pos:pos + 4]
        if tag == b'OBJT':
            return self._parse_objt(pos)
        elif tag == b'VOBJ':
            return self._parse_vobj(pos)
        return None, pos


def build_model_index(filepath):
    """Return {guid_lower: model_path} for every prototype that names a model."""
    db = ProtoDB(filepath)
    root = db.parse_root()
    idx = {}

    def walk(o):
        if isinstance(o, dict):
            g = o.get('GUID')
            if g:
                model = o.get('ModelName') or o.get('MarketModelName') or ''
                if model:
                    idx[g.lower()] = model.replace('\\', '/')
            for v in o.values():
                walk(v)
        elif isinstance(o, list):
            for v in o:
                walk(v)

    walk(root)
    return idx

File: mapeditor/src/protodb.cpp
// Native ProtoDB.bin -> {guid: model .srm path}. Same OBJT/VOBJ/SCHD field
// format as .map (see mapfile.cpp); here we walk EVERY object (incl. nested
// arrays) and collect GUID + ModelName. Port of cpcw_protodb.build_model_index.
#include "protodb.h"
#include <cstdint>
#include <cstring>
#include <fstream>
#include <vector>

namespace {

enum {
    FT_INT32=0x0001, FT_FLOAT=0x0002, FT_BOOL=0x0003, FT_STRING=0x0004,
    FT_FLOAT64=0x0005, FT_VEC3=0x0006, FT_GUID=0x0011, FT_REF=0x0012,
    FT_IID=0x0013, FT_ENTREF=0x0014, FT_VEC2F=0x0015, FT_VEC2I=0x0016,
    FT_UINT8=0x0017, FT_COLOR=0x0018, FT_INT16=0x0019, FT_LOCSTR=0x002B,
    FT_INLINE1=0x0088, FT_INLINE2=0x0089, FT_BLOB=0x0165, FT_FLAGS=0x039C,
    FT_ARRAY=0x898A
};
static bool stringLike(uint32_t t){ return t==FT_STRING||t==FT_GUID||t==FT_REF||t==FT_FLAGS||t==FT_LOCSTR; }

struct Field { std::string name; uint32_t type, size; };
struct Schema { std::string name; std::vector<Field> fields; };

struct DB {
    std::vector<uint8_t> b; size_t n=0;
    std::map<uint16_t, Schema> schemas;
    std::map<std::string,std::string>* out=nullptr;       // guid -> model (legacy)
    std::map<std::string,ProtoInfo>* full=nullptr;        // guid -> everything

    uint8_t  u8 (size_t p){ return p<n?b[p]:0; }
    uint16_t u16(size_t p){ return p+2<=n?(b[p]|(b[p+1]<<8)):0; }
    uint32_t u32(size_t p){ return p+4<=n?(uint32_t)(b[p]|(b[p+1]<<8)|(b[p+2]<<16)|((uint32_t)b[p+3]<<24)):0; }
    bool tag(size_t p,const char* t){ return p+4<=n && std::memcmp(b.data()+p,t,4)==0; }
    std::string str(size_t p,size_t& out){ uint16_t L=u16(p); size_t s=p+2; std::string r; if(s+L<=n)r.assign((const char*)b.data()+s,L); out=s+L; return r; }

    bool load(const std::string& path){
        std::ifstream f(path,std::ios::binary); if(!f) return false;
        f.seekg(0,std::ios::end); std::streamoff sz=f.tellg(); f.seekg(0);
        b.resize((size_t)sz); f.read((char*)b.data(),sz); n=b.size();
        if(!tag(0,"OBJS")) return false;
        parseSchd(u32(8));       // schema_offset @ +8
        return true;
    }
    void parseSchd(size_t pos){
        if(!tag(pos,"SCHD")) return;
        size_t limit=pos+8+u32(pos+4); pos+=8;
        uint16_t count=u16(pos); pos+=4;
        for(int s=0;s<count;s++){
            if(pos>=limit||!tag(pos,"SCHM")) break;
            size_t cs=pos+8, ce=cs+u32(pos+4), p=cs, np;
            Schema sc; sc.name=str(p,np); p=np;
            uint16_t typeId=u16(p); p+=2; p+=2; uint16_t fc=u16(p); p+=2; // type_id, version, field_count
            for(int fi=0;fi<fc;fi++){
                if(p+2>ce) break;
                Field fl; fl.name=str(p,np); p=np;
                if(p+8>ce) break;
                fl.type=u32(p); p+=4; fl.size=u32(p); p+=4;
                sc.fields.push_back(std::move(fl));
            }
            schemas[typeId]=std::move(sc);
            pos=ce;
        }
    }

    // field reader (advances correctly for every type; returns string for
    // string-like into `sv`, else sv stays empty)
    size_t readField(size_t p, uint32_t ft, uint32_t fs, size_t limit, std::string& sv){
        sv.clear();
        if(p>=limit) return p;
        if(ft==FT_INT32||ft==FT_FLOAT||ft==FT_COLOR||ft==FT_IID||ft==FT_ENTREF){ return p+4<=limit?p+4:limit; }
        if(ft==FT_BOOL||ft==FT_UINT8){ return p+1<=limit?p+1:limit; }
        if(ft==FT_INT16){ return p+2<=limit?p+2:limit; }
        if(ft==FT_FLOAT64){ return p+8<=limit?p+8:limit; }
        if(ft==FT_VEC3){ return p+12<=limit?p+12:limit; }
        if(ft==FT_VEC2F||ft==FT_VEC2I){ return p+8<=limit?p+8:limit; }
        if(stringLike(ft)){ if(p+2>limit)return limit; size_t np; sv=str(p,np); return np; }
        if(ft==FT_ARRAY) return readArray(p,limit);
        if(ft==FT_INLINE1||ft==FT_INLINE2) return readInline(p,limit);
        if(ft==FT_BLOB){ size_t nb=(fs>0&&fs<0xFFFF)?fs:0; return (nb>0&&p+nb<=limit)?p+nb:p; }
        if(fs>0&&fs<0xFFFF){ return p+fs<=limit?p+fs:p; }
        if(fs==0xFFFF||fs==0){
            if(tag(p,"ARRY")||tag(p,"OBJT")||tag(p,"VOBJ")) return readInline(p,limit);
            if(p+2<=limit){ uint16_t L=u16(p); if(L<4096&&p+2+L<=limit){ size_t np; sv=str(p,np); return np; } }
        }
        return p;
    }
    size_t readArray(size_t p, size_t limit){
        if(p+12>limit||!tag(p,"ARRY")) return p;
        size_t arrEnd=p+8+u32(p+4); uint32_t cnt=u32(p+8); p+=12;
        for(uint32_t k=0;k<cnt;k++){ if(p>=arrEnd)break; if(tag(p,"OBJT")){ size_t e; parseObjt(p,e); p=e; } else break; }
        return arrEnd;
    }
    struct Lift;
    size_t readInline(size_t p, size_t limit){
        if(tag(p,"OBJT")){ size_t e; parseObjt(p,e); return e; }
        if(tag(p,"VOBJ")){ size_t e; skipVobj(p,e); return e; }
        if(tag(p,"ARRY")) return readArray(p,limit);
        return p;
    }
    void skipVobj(size_t pos, size_t& endOut){
        endOut = tag(pos,"VOBJ") ? pos+8+u32(pos+4) : pos;
    }
    // Fields lifted out of one prototype object.
    struct Lift { std::string guid, model, name, schema; };

    void parseVobj(size_t pos, size_t& endOut, Lift& L){
        if(!tag(pos,"VOBJ")){ endOut=pos; return; }
        size_t contentEnd=pos+8+u32(pos+4); uint16_t typeId=u16(pos+8);
        auto it=schemas.find(typeId);
        size_t p=pos+10; endOut=contentEnd;
        if(p+2>contentEnd) return; p+=2;   // version
        if(it!=schemas.end()){
            if(L.schema.empty()) L.schema=it->second.name;
            for(auto& fl:it->second.fields){
                if(p>=contentEnd) break;
                std::string sv; p=readField(p,fl.type,fl.size,contentEnd,sv);
                if(!sv.empty()){
                    if(fl.name=="GUID") L.guid=sv;
                    else if(fl.name=="Name") { if(L.name.empty()) L.name=sv; }
                    else if(fl.name=="ModelName" || (L.model.empty()&&fl.name=="MarketModelName")) L.model=sv;
                }
            }
        }
    }
    void parseObjt(size_t pos, size_t& endOut){
        if(!tag(pos,"OBJT")){ endOut=pos; return; }
        size_t contentEnd=pos+8+u32(pos+4);
        Lift L; size_t ve; parseVobj(pos+10, ve, L);
        size_t p=ve;
        // trailing VOBJs extend the same object (derived schemas); merge them
        while(p+10<=contentEnd && tag(p,"VOBJ")){
            Lift x; size_t te; parseVobj(p,te,x); p=te;
            if(!x.guid.empty()) L.guid=x.guid;
            if(!x.name.empty() && L.name.empty()) L.name=x.name;
            if(!x.model.empty() && L.model.empty()) L.model=x.model;
        }
        if(!L.guid.empty()){
            std::string gl; for(char c:L.guid) gl+=(char)tolower((unsigned char)c);
            std::string mp; for(char c:L.model) mp+=(c=='\\'?'/':c);
            if(out && !mp.empty()) (*out)[gl]=mp;
            if(full){
                ProtoInfo& pi=(*full)[gl];
                if(pi.model.empty()) pi.model=mp;
                if(pi.name.empty())  pi.name=L.name;
                if(pi.schema.empty())pi.schema=L.schema;
            }
        }
        endOut=contentEnd;
    }
};

} // namespace

std::map<std::string,std::string> protodb_model_index(const std::string& path){
    std::map<std::string,std::string> idx;
    DB db; db.out=&idx;
    if(!db.load(path)) return idx;
    size_t e; db.parseObjt(12, e);   // root OBJT at offset 12
    return idx;
}

std::map<std::string,ProtoInfo> protodb_full_index(const std::string& path){
    std::map<std::string,ProtoInfo> idx;
    DB db; db.full=&idx;
    if(!db.load(path)) return idx;
    size_t e; db.parseObjt(12, e);
    return idx;
}

// ProtoDB prototype schemas and .map entity schemas are the same names in two
// dialects: "SP<X>" there, "S<X>Desc" here. SPDoodad -> SDoodadDesc,
// SPVehicleUnit -> SVehicleUnitDesc, SPBuildingUnit -> SBuildingUnitDesc.
std::string protodb_map_schema(const std::string& schema){
    if(schema.size()<3 || schema[0]!='S' || schema[1]!='P') return "";
    return "S" + schema.substr(2) + "Desc";
}
